sync_diff_inspector/diff.go

```diff
@@ -49,116 +51,266 @@ type Diff struct {
 }
 
 // NewDiff returns a Diff instance.
-func NewDiff(ctx context.Context, db1, db2 *sql.DB, cfg *Config) (diff *Diff, err error) {
+func NewDiff(ctx context.Context, cfg *Config) (diff *Diff, err error) {
 	diff = &Diff{
-		db1:              db1,
-		db2:              db2,
-		schema:           cfg.SourceDBCfg.Schema,
+		sourceDBs:        make(map[string]DBConfig),
 		chunkSize:        cfg.ChunkSize,
 		sample:           cfg.Sample,
 		checkThreadCount: cfg.CheckThreadCount,
 		useRowID:         cfg.UseRowID,
-		tables:           cfg.Tables,
+		useChecksum:      cfg.UseChecksum,
+		tables:           make(map[string]map[string]*TableConfig),
 		sqlCh:            make(chan string),
-		report:           NewReport(cfg.SourceDBCfg.Schema),
+		report:           NewReport(),
 		ctx:              ctx,
 	}
-	for _, table := range diff.tables {
-		table.Info, err = dbutil.GetTableInfoWithRowID(ctx, diff.db1, diff.schema, table.Name, cfg.UseRowID)
+
+	if err = diff.init(cfg); err != nil {
+		diff.Close()
+		return nil, errors.Trace(err)
+	}
+
+	return diff, nil
+}
+
+func (df *Diff) init(cfg *Config) (err error) {
+	// create connection for source.
+	for _, source := range cfg.SourceDBCfg {
+		source.Conn, err = dbutil.OpenDB(source.DBConfig)
 		if err != nil {
-			return nil, errors.Trace(err)
+			return errors.Errorf("create source db %+v error %v", cfg.SourceDBCfg, err)
+		}
+		source.Conn.SetMaxOpenConns(cfg.CheckThreadCount)
+		source.Conn.SetMaxIdleConns(cfg.CheckThreadCount)
```

Comment:
only control goroutines count is not enough? why need to control the underlying connection pool